File: api/routers/achievements.py
import logging
from fastapi import APIRouter, Depends, HTTPException
from api.services.achievement_service import AchievementService
from api.dependencies import get_current_user_id
from api.schemas.achievements import AchievementCheckResponse

logger = logging.getLogger(__name__)


def create_achievement_router(achievement_service: AchievementService) -> APIRouter:
    router = APIRouter()

    @router.get("/achievements")
    def get_user_achievements(user_id: int = Depends(get_current_user_id)):
        try:
            return achievement_service.get_user_achievements(user_id)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.post("/achievements/check", response_model=AchievementCheckResponse)
    def check_achievements(user_id: int = Depends(get_current_user_id)):
        try:
            new_achievements = achievement_service.check_and_award_achievements(user_id)
            return {"new_achievements": new_achievements, "count": len(new_achievements)}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/achievements/progress")
    def achievements_progress(user_id: int = Depends(get_current_user_id)):
        try:
            return achievement_service.db.get_achievements_progress(user_id)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    return router

File: functions/_shared/routes/achievements.ts
import { Hono } from 'hono';
import type { AppEnv } from '../lib/env';
import { createDb } from '../db/client';
import {
  ACHIEVEMENT_METADATA,
  checkAchievements,
  getAchievementsProgress,
  getUserAchievements,
} from '../services/achievement-service';

export const achievementRoutes = new Hono<AppEnv>();

function withMetadata(achievementType: string) {
  return ACHIEVEMENT_METADATA[achievementType] ?? {};
}

achievementRoutes.get('/achievements', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);

  const achievements = await getUserAchievements(db, userId);
  return c.json(achievements.map((achievement) => ({ ...achievement, ...withMetadata(achievement.achievement_type) })));
});

achievementRoutes.post('/achievements/check', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);

  const newAchievementTypes = await checkAchievements(db, userId);
  const newAchievements = newAchievementTypes.map((achievementType) => ({
    ...withMetadata(achievementType),
    achievement_type: achievementType,
  }));

  return c.json({ new_achievements: newAchievements, count: newAchievements.length });
});

achievementRoutes.get('/achievements/progress', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);
  return c.json(await getAchievementsProgress(db, userId));
});
